src/ReservoirPriceOracle.sol

```diff
@@ -1,32 +1,405 @@
+import { ReentrancyGuard } from "lib/amm-core/lib/solmate/src/utils/ReentrancyGuard.sol";
+import { FixedPointMathLib } from "lib/amm-core/lib/solady/src/utils/FixedPointMathLib.sol";
 
-contract ReservoirPriceOracle is IReservoirPriceOracle {
-    function getTimeWeightedAverage(OracleAverageQuery[] memory aQueries)
+contract ReservoirPriceOracle is IPriceOracle, IReservoirPriceOracle, Owned(msg.sender), ReentrancyGuard {
+    using FixedPointMathLib for uint256;
+    using QueryProcessor for ReservoirPair;
+    using Utils for *;
+
+    ///////////////////////////////////////////////////////////////////////////////////////////////
+    //                                       CONSTANTS                                           //
+    ///////////////////////////////////////////////////////////////////////////////////////////////
+
+    uint256 private constant MAX_DEVIATION_THRESHOLD = 0.1e18; // 10%
+    uint256 private constant MAX_TWAP_PERIOD = 1 hours;
+    uint256 private constant MAX_ROUTE_LENGTH = 4;
+    uint256 private constant WAD = 1e18;
+
+    ///////////////////////////////////////////////////////////////////////////////////////////////
+    //                                       EVENTS                                              //
+    ///////////////////////////////////////////////////////////////////////////////////////////////
+
+    event DesignatePair(address token0, address token1, ReservoirPair pair);
+    event PriceDeviationThreshold(uint256 newThreshold);
+    event RewardMultiplier(uint256 newMultiplier);
+    event Route(address token0, address token1, address[] route);
+    event Price(address token0, address token1, uint256 price);
+    event TwapPeriod(uint256 newPeriod);
+
+    ///////////////////////////////////////////////////////////////////////////////////////////////
+    //                                       ERRORS                                              //
+    ///////////////////////////////////////////////////////////////////////////////////////////////
+
+    error RPC_THRESHOLD_TOO_HIGH();
+    error RPC_INVALID_TWAP_PERIOD();
+    error RPC_SAME_TOKEN();
+    error RPC_TOKENS_UNSORTED();
+    error RPC_INVALID_ROUTE_LENGTH();
+    error RPC_INVALID_ROUTE();
+    error RPC_UPDATE_PRICE_NOT_SIMPLE_ROUTE();
+
+    ///////////////////////////////////////////////////////////////////////////////////////////////
+    //                                        STORAGE                                            //
+    ///////////////////////////////////////////////////////////////////////////////////////////////
+
+    /// @notice percentage change greater than which, a price update with the oracles would succeed
+    /// 1e18 == 100%
+    uint64 public priceDeviationThreshold;
+
+    /// @notice multiples of the base fee the contract rewards the caller for updating the price when it goes
+    /// beyond the `priceDeviationThreshold`
+    uint64 public rewardMultiplier;
```

Comment:
Should this not just be a gas amount that is then multiplied by base fee, or how do you compute the amount of gas to pay out as a reward?